### PR TITLE
executor: support generated column on point get and batch point get

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve?

Problem Summary:

To make the lock behaviors the same for accessing unique index with general columns and common columns.

### What is changed and how it works?

Let the point get and batch point get could be used for accessing data using an index on virtual generated column

### Check List 

Tests 

- Unit test

Side effects

### CODE REVIEW

File: planner/core/physical_plans.go [hunk 1]
```diff
@@ -260,17 +260,18 @@ func (ts *PhysicalTableScan) IsPartition() (bool, int64) {
 }
 
 // ExpandVirtualColumn expands the virtual column's dependent columns to ts's schema and column.
-func (ts *PhysicalTableScan) ExpandVirtualColumn() {
-	for _, col := range ts.schema.Columns {
+func ExpandVirtualColumn(columns *[]*model.ColumnInfo, schema *expression.Schema, colsInfo []*model.ColumnInfo) {
```
Comment (coocood) [inline planner/core/physical_plans.go]: Returning the columns is more idiomatic.

File: session/pessimistic_test.go [hunk 2]
```diff
@@ -1162,3 +1162,37 @@ func (s *testPessimisticSuite) TestRCSubQuery(c *C) {
 	tk.MustQuery("select * from t1 where c1 = (select c1 from t where c1 = 1) and 1=1;").Check(testkit.Rows("1 4"))
 	tk.MustExec("rollback")
 }
+
+func (s *testPessimisticSuite) TestGenerateColPointGet(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
```
Comment (lysu) [inline session/pessimistic_test.go]: how about help a test case

```
set global tidb_row_format_version=1;
````
then re-run those script to ensure it also work for old row format

**Comment by coocood** [comment]
LGTM

File: planner/core/point_get_plan.go [hunk 3]
```diff
@@ -165,6 +167,10 @@ func (p *PointGetPlan) SetChild(i int, child PhysicalPlan) {}
 
 // ResolveIndices resolves the indices for columns. After doing this, the columns can evaluate the rows by their indices.
 func (p *PointGetPlan) ResolveIndices() error {
+	err := resolveIndicesForVirtualColumn(p.schema.Columns, p.schema)
+	if err != nil {
+		return err
+	}
 	return nil
```
Comment (Deardrops) [inline planner/core/point_get_plan.go]:
```suggestion
	return resolveIndicesForVirtualColumn(p.schema.Columns, p.schema)
```

File: planner/core/point_get_plan.go [hunk 4]
```diff
@@ -290,6 +297,10 @@ func (p *BatchPointGetPlan) SetChild(i int, child PhysicalPlan) {}
 
 // ResolveIndices resolves the indices for columns. After doing this, the columns can evaluate the rows by their indices.
 func (p *BatchPointGetPlan) ResolveIndices() error {
+	err := resolveIndicesForVirtualColumn(p.schema.Columns, p.schema)
+	if err != nil {
+		return err
+	}
 	return nil
```
Comment (Deardrops) [inline planner/core/point_get_plan.go]: ditto

File: session/pessimistic_test.go [hunk 5]
```diff
@@ -1162,3 +1163,44 @@ func (s *testPessimisticSuite) TestRCSubQuery(c *C) {
...
+		tk.MustExec("CREATE TABLE `tu`(`x` int, `y` int, `z` int GENERATED ALWAYS AS (x + y) VIRTUAL, PRIMARY KEY (`x`), UNIQUE KEY `idz` (`z`))")
+		tk.MustExec("insert into tu(x, y) values(1, 2);")
+
+		// test point get lock
+		tk.MustExec("begin pessimistic")
+		tk.MustQuery("select * from tu where z = 3 for update").Check(testkit.Rows("1 2 3"))
+		tk2 := testkit.NewTestKitWithInit(c, s.store)
+		tk2.MustExec("begin pessimistic")
+		err := tk2.ExecToErr("select * from tu where z = 3 for update nowait")
+		c.Assert(err, NotNil)
```
Comment (Deardrops) [inline session/pessimistic_test.go]: Could we assert the error message here?

File: executor/builder.go [hunk 6]
```diff
@@ -2910,14 +2910,19 @@ func newRowDecoder(ctx sessionctx.Context, schema *expression.Schema, tbl *model
...
-			Elems:   col.RetType.Elems,
-			Collate: col.GetType().Collate,
+			ID:            col.ID,
+			Tp:            int32(col.RetType.Tp),
+			Flag:          int32(col.RetType.Flag),
+			Flen:          col.RetType.Flen,
+			Decimal:       col.RetType.Decimal,
+			Elems:         col.RetType.Elems,
+			Collate:       col.GetType().Collate,
+			VirtualGenCol: isGeneratedCol,
```
Comment (wjhuang2016) [inline executor/builder.go]: `col.VirtualExpr != nil` is enough?

Comment (cfzjywxk) [reply executor/builder.go]: Some exceptions should be taken into account?

**Comment by Deardrops** [comment]
LGTM

**Comment by jackysp** [comment]
PTAL @wjhuang2016 @lysu

File: executor/builder.go (hunk 6 above)
Comment (wjhuang2016) [reply executor/builder.go]: No, just for the simpleness.

**Pull request merged**